api.R

```diff
@@ -1,101 +1,238 @@
+  training_data <- read_parquet(
+    file.path(
+      dvc_bucket,
+      substr(training_data_md5, 1, 2),
+      substr(training_data_md5, 3, nchar(training_data_md5))
+    )
+  )
 
-# Load Model -------------------------------------------------------------------
 
-# Load fit and recipe from file
-fit <- lightsnip::lgbm_load(temp_file_fit)
-recipe <- readRDS(temp_file_recipe)
+  # Load Model -----------------------------------------------------------------
 
-# Extract a sample row of predictors to use for the API docs
-predictors <- recipe$var_info %>%
-  filter(role == "predictor") %>%
-  pull(variable)
-ptype_tbl <- training_data %>%
-  filter(meta_pin == "15251030220000") %>%
-  select(all_of(predictors))
-ptype <- vetiver_create_ptype(model = fit, save_prototype = ptype_tbl)
+  # Load fit and recipe from file
+  fit <- lightsnip::lgbm_load(temp_file_fit)
+  recipe <- readRDS(temp_file_recipe)
 
+  # Extract a sample row of data to use for the API docs
+  ptype_tbl <- training_data %>%
+    filter(meta_pin == "15251030220000")
 
-# Create API -------------------------------------------------------------------
+  # If the model recipe is configured to allow it, strip all chars except
+  # for the predictors from the example row
+  if (predictors_only) {
+    predictors <- recipe$var_info %>%
+      filter(role == "predictor") %>%
+      pull(variable)
+    ptype_tbl <- ptype_tbl %>%
+      filter(meta_pin == "15251030220000") %>%
+      select(all_of(predictors))
+  }
 
-# Create model object and populate metadata
-model <- vetiver_model(fit, "LightGBM", save_prototype = ptype)
-model$recipe <- recipe
-model$pv$round_type <- metadata$pv_round_type
-model$pv$round_break <- metadata$pv_round_break[[1]]
-model$pv$round_to_nearest <- metadata$pv_round_to_nearest[[1]]
+  ptype <- vetiver_create_ptype(model = fit, save_prototype = ptype_tbl)
 
-# Start API
-pr() %>%
-  vetiver_api(model) %>%
-  pr_run(
-    host = "0.0.0.0",
-    port = api_port
+
+  # Create API -----------------------------------------------------------------
+
+  # Create model object and populate metadata
+  model <- vetiver_model(fit, "LightGBM", save_prototype = ptype)
+  model$recipe <- recipe
+  model$pv$round_type <- metadata$pv_round_type
+  model$pv$round_break <- metadata$pv_round_break[[1]]
+  model$pv$round_to_nearest <- metadata$pv_round_to_nearest[[1]]
+
+  return(model)
+}
+
+# Filter the valid runs for the run marked as default
+default_run <- valid_runs %>%
+  dplyr::filter(run_id == default_run_id) %>%
+  dplyr::slice_head()
+
+# Retrieve paths and model objects for all endpoints to be deployed
+all_endpoints <- list()
+for (i in seq_len(nrow(valid_runs))) {
+  run <- valid_runs[i, ]
+  model <- get_model_from_run(
+    run$run_id, run$year, run$dvc_bucket, run$predictors_only
   )
+  all_endpoints[[i]] <- list(
+    path = glue::glue("{base_url_prefix}/{run$run_id}"),
+    model = model,
+    is_default = run$run_id == default_run$run_id
```

Comment:
In a similar vein, it would make a lot more sense to just append another entry to `all_endpoints` for the default run rather than have to check the `is_default` bool in all of the iteration blocks that follow, but I couldn't figure out a good way to do this given the way append operations in R seem to require index references (i.e. the `all_endpoints[[i]]` assignment on line 167).

Reply:
You should be able to use

```r
all_endpoints <- c(all_endpoints, default_endpoint)`
```
or:

```r
all_endpoints <- append(all_endpoints, default_endpoint)`
```

If that simplifies things.

Reply:
Very nice, that's much clearer! I made this change in [2a770c0](https://github.com/ccao-data/api-res-avm/pull/2/commits/2a770c030e7e37dbbf20c98940e8183040012307). Sadly I needed to upgrade the entire set of dependencies in order to get the environment working with R 4.4.0 and test this, which blew up the diff in [bc8c7f5](https://github.com/ccao-data/api-res-avm/pull/2/commits/bc8c7f5fc14e6f8c1d18eb010e3714b29de9fea3); I'll go in and comment the extra changes I made to make it a bit clearer.